File: frontend/src/pages/advanced/namespace-comparison/utils/comparison.ts
import type { NamespaceSnapshot, CompareRow } from '../types/index'
import {
  getCapacitySeverity,
  getStabilitySeverity,
  getExposureSeverity,
  getQuotaSeverity,
  getStorageSeverity,
  getRBACSeverity
} from './compute'

// Format bytes to human readable
function formatBytes(bytes: number): string {
  if (bytes === 0) return '0 B'
  const k = 1024
  const sizes = ['B', 'Ki', 'Mi', 'Gi', 'Ti']
  const i = Math.floor(Math.log(bytes) / Math.log(k))
  return `${(bytes / Math.pow(k, i)).toFixed(2)} ${sizes[i]}`
}

// Format number with suffix
function formatNumber(num: number): string {
  if (num >= 1000000) return `${(num / 1000000).toFixed(1)}M`
  if (num >= 1000) return `${(num / 1000).toFixed(1)}K`
  return num.toString()
}

// Format percentage
function formatPercentage(value?: number): string {
  if (value === undefined) return '–'
  return `${value.toFixed(1)}%`
}

// Calculate delta between two values
function calculateDelta(
  valueA: string | number,
  valueB: string | number
): string | number | undefined {
  if (typeof valueA === 'number' && typeof valueB === 'number') {
    const delta = valueB - valueA
    if (delta === 0) return undefined
    return delta > 0 ? `+${delta}` : delta.toString()
  }
  
  if (valueA === valueB) return undefined
  
  // For boolean-like values
  if (valueA === 'Yes' && valueB === 'No') return '↓'
  if (valueA === 'No' && valueB === 'Yes') return '↑'
  
  return undefined
}

// Compare two namespace snapshots
export function compareSnapshots(
  snapshotA: NamespaceSnapshot,
  snapshotB: NamespaceSnapshot
): CompareRow[] {
  const rows: CompareRow[] = []
  
  // Section: Capacity at a Glance
  rows.push({ section: 'Capacity at a Glance', metric: '', valueA: '', valueB: '' })
  
  // CPU
  rows.push({
    metric: 'CPU Requests',
    valueA: formatNumber(snapshotA.capacity.cpu.requests),
    valueB: formatNumber(snapshotB.capacity.cpu.requests),
    delta: calculateDelta(snapshotA.capacity.cpu.requests, snapshotB.capacity.cpu.requests),
    tooltip: 'Total CPU cores requested by all pods'
  })
  
  rows.push({
    metric: 'CPU Usage',
    valueA: snapshotA.capacity.cpu.usage !== undefined 
      ? formatNumber(snapshotA.capacity.cpu.usage) 
      : '–',
    valueB: snapshotB.capacity.cpu.usage !== undefined 
      ? formatNumber(snapshotB.capacity.cpu.usage) 
      : '–',
    delta: snapshotA.capacity.cpu.usage !== undefined && snapshotB.capacity.cpu.usage !== undefined
      ? calculateDelta(snapshotA.capacity.cpu.usage, snapshotB.capacity.cpu.usage)
      : undefined,
    tooltip: 'Actual CPU usage from metrics'
  })
  
  rows.push({
    metric: 'CPU Headroom',
    valueA: formatPercentage(snapshotA.capacity.cpu.headroom),
    valueB: formatPercentage(snapshotB.capacity.cpu.headroom),
    severityA: getCapacitySeverity(snapshotA.capacity.cpu.headroom),
    severityB: getCapacitySeverity(snapshotB.capacity.cpu.headroom),
    tooltip: 'Available CPU capacity (requests - usage) / requests'
  })
  
  // Memory
  rows.push({
    metric: 'Memory Requests',
    valueA: formatBytes(snapshotA.capacity.memory.requests),
    valueB: formatBytes(snapshotB.capacity.memory.requests),
    delta: calculateDelta(snapshotA.capacity.memory.requests, snapshotB.capacity.memory.requests),
    tooltip: 'Total memory requested by all pods'
  })
  
  rows.push({
    metric: 'Memory Usage',
    valueA: snapshotA.capacity.memory.usage !== undefined 
      ? formatBytes(snapshotA.capacity.memory.usage) 
      : '–',
    valueB: snapshotB.capacity.memory.usage !== undefined 
      ? formatBytes(snapshotB.capacity.memory.usage) 
      : '–',
    delta: snapshotA.capacity.memory.usage !== undefined && snapshotB.capacity.memory.usage !== undefined
      ? calculateDelta(snapshotA.capacity.memory.usage, snapshotB.capacity.memory.usage)
      : undefined,
    tooltip: 'Actual memory usage from metrics'
  })
  
  rows.push({
    metric: 'Memory Headroom',
    valueA: formatPercentage(snapshotA.capacity.memory.headroom),
    valueB: formatPercentage(snapshotB.capacity.memory.headroom),
    severityA: getCapacitySeverity(snapshotA.capacity.memory.headroom),
    severityB: getCapacitySeverity(snapshotB.capacity.memory.headroom),
    tooltip: 'Available memory capacity (requests - usage) / requests'
  })
  
  // Section: Pod Stability
  rows.push({ section: 'Pod Stability', metric: '', valueA: '', valueB: '' })
  
  rows.push({
    metric: 'CrashLoops (24h)',
    valueA: snapshotA.stability.crashLoops24h,
    valueB: snapshotB.stability.crashLoops24h,
    delta: calculateDelta(snapshotA.stability.crashLoops24h, snapshotB.stability.crashLoops24h),
    severityA: getStabilitySeverity('crashLoops', snapshotA.stability.crashLoops24h),
    severityB: getStabilitySeverity('crashLoops', snapshotB.stability.crashLoops24h),
    tooltip: 'Pods in CrashLoopBackOff state in the last 24 hours'
  })
  
  rows.push({
    metric: 'Restarts (24h)',
    valueA: snapshotA.stability.restarts24h,
    valueB: snapshotB.stability.restarts24h,
    delta: calculateDelta(snapshotA.stability.restarts24h, snapshotB.stability.restarts24h),
    severityA: getStabilitySeverity('restarts', snapshotA.stability.restarts24h),
    severityB: getStabilitySeverity('restarts', snapshotB.stability.restarts24h),
    tooltip: 'Total container restarts in the last 24 hours'
  })
  
  rows.push({
    metric: 'Pending >5m',
    valueA: snapshotA.stability.pendingOver5m,
    valueB: snapshotB.stability.pendingOver5m,
    delta: calculateDelta(snapshotA.stability.pendingOver5m, snapshotB.stability.pendingOver5m),
    severityA: getStabilitySeverity('pending', snapshotA.stability.pendingOver5m),
    severityB: getStabilitySeverity('pending', snapshotB.stability.pendingOver5m),
    tooltip: 'Pods stuck in Pending state for more than 5 minutes'
  })
  
  // Section: Exposure Surface
  rows.push({ section: 'Exposure Surface', metric: '', valueA: '', valueB: '' })
  
  rows.push({
    metric: 'LoadBalancer Services',
    valueA: snapshotA.exposure.services.loadBalancer,
    valueB: snapshotB.exposure.services.loadBalancer,
    delta: calculateDelta(
      snapshotA.exposure.services.loadBalancer,
      snapshotB.exposure.services.loadBalancer
    ),
    tooltip: 'Services exposed via LoadBalancer'
  })
  
  rows.push({
    metric: 'NodePort Services',
    valueA: snapshotA.exposure.services.nodePort,
    valueB: snapshotB.exposure.services.nodePort,
    delta: calculateDelta(
      snapshotA.exposure.services.nodePort,
      snapshotB.exposure.services.nodePort
    ),
    tooltip: 'Services exposed via NodePort'
  })
  
  rows.push({
    metric: 'ClusterIP Services',
    valueA: snapshotA.exposure.services.clusterIP,
    valueB: snapshotB.exposure.services.clusterIP,
    delta: calculateDelta(
      snapshotA.exposure.services.clusterIP,
      snapshotB.exposure.services.clusterIP
    ),
    tooltip: 'Internal ClusterIP services'
  })
  
  rows.push({
    metric: 'Ingress Hosts',
    valueA: snapshotA.exposure.ingressHosts,
    valueB: snapshotB.exposure.ingressHosts,
    delta: calculateDelta(snapshotA.exposure.ingressHosts, snapshotB.exposure.ingressHosts),
    tooltip: 'Unique hostnames in Ingress rules'
  })
  
  rows.push({
    metric: 'NetworkPolicy Present',
    valueA: snapshotA.exposure.hasNetworkPolicy ? 'Yes' : 'No',
    valueB: snapshotB.exposure.hasNetworkPolicy ? 'Yes' : 'No',
    delta: calculateDelta(
      snapshotA.exposure.hasNetworkPolicy ? 'Yes' : 'No',
      snapshotB.exposure.hasNetworkPolicy ? 'Yes' : 'No'
    ),
    severityA: getExposureSeverity(snapshotA.exposure.services, snapshotA.exposure.hasNetworkPolicy),
    severityB: getExposureSeverity(snapshotB.exposure.services, snapshotB.exposure.hasNetworkPolicy),
    tooltip: 'Whether any NetworkPolicy is defined'
  })
  
  // Section: Quota & Defaults Hygiene
  rows.push({ section: 'Quota & Defaults Hygiene', metric: '', valueA: '', valueB: '' })
  
  rows.push({
    metric: 'ResourceQuota Present',
    valueA: snapshotA.quota.hasResourceQuota ? 'Yes' : 'No',
    valueB: snapshotB.quota.hasResourceQuota ? 'Yes' : 'No',
    delta: calculateDelta(
      snapshotA.quota.hasResourceQuota ? 'Yes' : 'No',
      snapshotB.quota.hasResourceQuota ? 'Yes' : 'No'
    ),
    tooltip: 'Whether ResourceQuota is defined'
  })
  
  // Top quota usage
  if (snapshotA.quota.topUsage.length > 0 || snapshotB.quota.topUsage.length > 0) {
    const maxItems = Math.max(snapshotA.quota.topUsage.length, snapshotB.quota.topUsage.length)
    
    for (let i = 0; i < Math.min(maxItems, 3); i++) {
      const usageA = snapshotA.quota.topUsage[i]
      const usageB = snapshotB.quota.topUsage[i]
      
      rows.push({
        metric: `  Top Usage ${i + 1}`,
        valueA: usageA 
          ? `${usageA.resource}: ${usageA.used}/${usageA.hard} (${usageA.percentage.toFixed(0)}%)`
          : '–',
        valueB: usageB 
          ? `${usageB.resource}: ${usageB.used}/${usageB.hard} (${usageB.percentage.toFixed(0)}%)`
          : '–',
        severityA: usageA ? (usageA.percentage >= 95 ? 'crit' : usageA.percentage >= 80 ? 'warn' : 'ok') : 'ok',
        severityB: usageB ? (usageB.percentage >= 95 ? 'crit' : usageB.percentage >= 80 ? 'warn' : 'ok') : 'ok',
        tooltip: 'Resource quota usage by percentage'
      })
    }
  }
  
  rows.push({
    metric: 'LimitRange Present',
    valueA: snapshotA.quota.hasLimitRange ? 'Yes' : 'No',
    valueB: snapshotB.quota.hasLimitRange ? 'Yes' : 'No',
    delta: calculateDelta(
      snapshotA.quota.hasLimitRange ? 'Yes' : 'No',
      snapshotB.quota.hasLimitRange ? 'Yes' : 'No'
    ),
    tooltip: 'Whether LimitRange is defined'
  })
  
  // Section: Storage Basics
  rows.push({ section: 'Storage Basics', metric: '', valueA: '', valueB: '' })
  
  rows.push({
    metric: 'PVC Count',
    valueA: snapshotA.storage.pvcCount,
    valueB: snapshotB.storage.pvcCount,
    delta: calculateDelta(snapshotA.storage.pvcCount, snapshotB.storage.pvcCount),
    tooltip: 'Total number of PersistentVolumeClaims'
  })
  
  rows.push({
    metric: 'Requested Storage',
    valueA: `${snapshotA.storage.requestedGi.toFixed(1)} Gi`,
    valueB: `${snapshotB.storage.requestedGi.toFixed(1)} Gi`,
    delta: calculateDelta(snapshotA.storage.requestedGi, snapshotB.storage.requestedGi),
    tooltip: 'Total storage requested in GiB'
  })
  
  rows.push({
    metric: 'Unbound PVCs',
    valueA: snapshotA.storage.unboundPVCs,
    valueB: snapshotB.storage.unboundPVCs,
    delta: calculateDelta(snapshotA.storage.unboundPVCs, snapshotB.storage.unboundPVCs),
    severityA: getStorageSeverity(snapshotA.storage.unboundPVCs, snapshotA.storage.orphanedPVCs),
    severityB: getStorageSeverity(snapshotB.storage.unboundPVCs, snapshotB.storage.orphanedPVCs),
    tooltip: 'PVCs not bound to a PersistentVolume'
  })
  
  rows.push({
    metric: 'Orphaned PVCs',
    valueA: snapshotA.storage.orphanedPVCs,
    valueB: snapshotB.storage.orphanedPVCs,
    delta: calculateDelta(snapshotA.storage.orphanedPVCs, snapshotB.storage.orphanedPVCs),
    severityA: getStorageSeverity(snapshotA.storage.unboundPVCs, snapshotA.storage.orphanedPVCs),
    severityB: getStorageSeverity(snapshotB.storage.unboundPVCs, snapshotB.storage.orphanedPVCs),
    tooltip: 'PVCs not mounted by any pod'
  })
  
  // Section: RBAC Risk Snapshot
  rows.push({ section: 'RBAC Risk Snapshot', metric: '', valueA: '', valueB: '' })
  
  rows.push({
    metric: 'Admin Bindings',
    valueA: snapshotA.rbac.adminBindings,
    valueB: snapshotB.rbac.adminBindings,
    delta: calculateDelta(snapshotA.rbac.adminBindings, snapshotB.rbac.adminBindings),
    severityA: getRBACSeverity(snapshotA.rbac.adminBindings, snapshotA.rbac.wildcardRules),
    severityB: getRBACSeverity(snapshotB.rbac.adminBindings, snapshotB.rbac.wildcardRules),
    tooltip: 'RoleBindings to cluster-admin, admin, or edit ClusterRoles'
  })
  
  rows.push({
    metric: 'Wildcard Rules',
    valueA: snapshotA.rbac.wildcardRules,
    valueB: snapshotB.rbac.wildcardRules,
    delta: calculateDelta(snapshotA.rbac.wildcardRules, snapshotB.rbac.wildcardRules),
    severityA: getRBACSeverity(snapshotA.rbac.adminBindings, snapshotA.rbac.wildcardRules),
    severityB: getRBACSeverity(snapshotB.rbac.adminBindings, snapshotB.rbac.wildcardRules),
    tooltip: 'Rules with * in resources or verbs'
  })
  
  return rows
}

File: backend/internal/api/handlers/namespaces/comparison.go
package namespaces

import (
	"fmt"
	"math"
	"strconv"
	"strings"
)

// Comparator compares two namespace snapshots
type Comparator struct{}

// NewComparator creates a new namespace comparator
func NewComparator() *Comparator {
	return &Comparator{}
}

// Compare generates comparison rows between two snapshots
func (c *Comparator) Compare(snapshotA, snapshotB *NamespaceSnapshot) []CompareRow {
	var rows []CompareRow
	
	// Compare capacity metrics
	rows = append(rows, c.compareCapacity(snapshotA.Capacity, snapshotB.Capacity)...)
	
	// Compare stability metrics
	rows = append(rows, c.compareStability(snapshotA.Stability, snapshotB.Stability)...)
	
	// Compare exposure metrics
	rows = append(rows, c.compareExposure(snapshotA.Exposure, snapshotB.Exposure)...)
	
	// Compare quota metrics
	rows = append(rows, c.compareQuota(snapshotA.Quota, snapshotB.Quota)...)
	
	// Compare storage metrics
	rows = append(rows, c.compareStorage(snapshotA.Storage, snapshotB.Storage)...)
	
	// Compare RBAC metrics
	rows = append(rows, c.compareRBAC(snapshotA.RBAC, snapshotB.RBAC)...)
	
	return rows
}

func (c *Comparator) compareCapacity(a, b CapacityMetrics) []CompareRow {
	rows := []CompareRow{
		{
			Category:    "Capacity",
			Metric:      "CPU Requested",
			ValueA:      a.CPU.Requested,
			ValueB:      b.CPU.Requested,
			Severity:    c.calculateResourceSeverity(a.CPU.Requested, b.CPU.Requested),
			Trend:       c.calculateTrend(a.CPU.Requested, b.CPU.Requested),
			Significant: c.isSignificantDifference(a.CPU.Requested, b.CPU.Requested, 0.2),
		},
		{
			Category:    "Capacity",
			Metric:      "CPU Limit",
			ValueA:      a.CPU.Limit,
			ValueB:      b.CPU.Limit,
			Severity:    c.calculateResourceSeverity(a.CPU.Limit, b.CPU.Limit),
			Trend:       c.calculateTrend(a.CPU.Limit, b.CPU.Limit),
			Significant: c.isSignificantDifference(a.CPU.Limit, b.CPU.Limit, 0.2),
		},
		{
			Category:    "Capacity",
			Metric:      "CPU Headroom",
			ValueA:      fmt.Sprintf("%.1f%%", a.CPU.Headroom),
			ValueB:      fmt.Sprintf("%.1f%%", b.CPU.Headroom),
			Severity:    c.calculateHeadroomSeverity(a.CPU.Headroom, b.CPU.Headroom),
			Trend:       c.calculateFloatTrend(a.CPU.Headroom, b.CPU.Headroom),
			Significant: math.Abs(a.CPU.Headroom-b.CPU.Headroom) > 20,
		},
		{
			Category:    "Capacity",
			Metric:      "Memory Requested",
			ValueA:      a.Memory.Requested,
			ValueB:      b.Memory.Requested,
			Severity:    c.calculateResourceSeverity(a.Memory.Requested, b.Memory.Requested),
			Trend:       c.calculateTrend(a.Memory.Requested, b.Memory.Requested),
			Significant: c.isSignificantDifference(a.Memory.Requested, b.Memory.Requested, 0.2),
		},
		{
			Category:    "Capacity",
			Metric:      "Memory Limit",
			ValueA:      a.Memory.Limit,
			ValueB:      b.Memory.Limit,
			Severity:    c.calculateResourceSeverity(a.Memory.Limit, b.Memory.Limit),
			Trend:       c.calculateTrend(a.Memory.Limit, b.Memory.Limit),
			Significant: c.isSignificantDifference(a.Memory.Limit, b.Memory.Limit, 0.2),
		},
		{
			Category:    "Capacity",
			Metric:      "Memory Headroom",
			ValueA:      fmt.Sprintf("%.1f%%", a.Memory.Headroom),
			ValueB:      fmt.Sprintf("%.1f%%", b.Memory.Headroom),
			Severity:    c.calculateHeadroomSeverity(a.Memory.Headroom, b.Memory.Headroom),
			Trend:       c.calculateFloatTrend(a.Memory.Headroom, b.Memory.Headroom),
			Significant: math.Abs(a.Memory.Headroom-b.Memory.Headroom) > 20,
		},
	}
	
	return rows
}

func (c *Comparator) compareStability(a, b StabilityMetrics) []CompareRow {
	rows := []CompareRow{
		{
			Category:    "Stability",
			Metric:      "Crash Loops (24h)",
			ValueA:      a.CrashLoops24h,
			ValueB:      b.CrashLoops24h,
			Severity:    c.calculateStabilitySeverity(a.CrashLoops24h, b.CrashLoops24h),
			Trend:       c.calculateIntTrend(a.CrashLoops24h, b.CrashLoops24h),
			Significant: a.CrashLoops24h != b.CrashLoops24h,
		},
		{
			Category:    "Stability",
			Metric:      "Restarts (24h)",
			ValueA:      a.Restarts24h,
			ValueB:      b.Restarts24h,
			Severity:    c.calculateStabilitySeverity(a.Restarts24h, b.Restarts24h),
			Trend:       c.calculateIntTrend(a.Restarts24h, b.Restarts24h),
			Significant: math.Abs(float64(a.Restarts24h-b.Restarts24h)) > 5,
		},
		{
			Category:    "Stability",
			Metric:      "Pending Pods (>5m)",
			ValueA:      a.PendingOver5m,
			ValueB:      b.PendingOver5m,
			Severity:    c.calculatePendingSeverity(a.PendingOver5m, b.PendingOver5m),
			Trend:       c.calculateIntTrend(a.PendingOver5m, b.PendingOver5m),
			Significant: a.PendingOver5m != b.PendingOver5m,
		},
	}
	
	return rows
}

func (c *Comparator) compareExposure(a, b ExposureMetrics) []CompareRow {
	rows := []CompareRow{
		{
			Category:    "Exposure",
			Metric:      "Services",
			ValueA:      a.Services,
			ValueB:      b.Services,
			Severity:    "normal",
			Trend:       c.calculateIntTrend(a.Services, b.Services),
			Significant: math.Abs(float64(a.Services-b.Services)) > 3,
		},
		{
			Category:    "Exposure",
			Metric:      "Ingress Hosts",
			ValueA:      len(a.IngressHosts),
			ValueB:      len(b.IngressHosts),
			Severity:    c.calculateExposureSeverity(len(a.IngressHosts), len(b.IngressHosts)),
			Trend:       c.calculateIntTrend(len(a.IngressHosts), len(b.IngressHosts)),
			Significant: len(a.IngressHosts) != len(b.IngressHosts),
		},
		{
			Category:    "Exposure",
			Metric:      "Network Policy",
			ValueA:      c.boolToString(a.HasNetworkPolicy),
			ValueB:      c.boolToString(b.HasNetworkPolicy),
			Severity:    c.calculateNetPolSeverity(a.HasNetworkPolicy, b.HasNetworkPolicy),
			Trend:       "equal",
			Significant: a.HasNetworkPolicy != b.HasNetworkPolicy,
		},
	}
	
	return rows
}

func (c *Comparator) compareQuota(a, b QuotaMetrics) []CompareRow {
	rows := []CompareRow{
		{
			Category:    "Quota",
			Metric:      "Resource Quota",
			ValueA:      c.boolToString(a.HasResourceQuota),
			ValueB:      c.boolToString(b.HasResourceQuota),
			Severity:    c.calculateQuotaSeverity(a.HasResourceQuota, b.HasResourceQuota),
			Trend:       "equal",
			Significant: a.HasResourceQuota != b.HasResourceQuota,
		},
		{
			Category:    "Quota",
			Metric:      "Limit Range",
			ValueA:      c.boolToString(a.HasLimitRange),
			ValueB:      c.boolToString(b.HasLimitRange),
			Severity:    c.calculateQuotaSeverity(a.HasLimitRange, b.HasLimitRange),
			Trend:       "equal",
			Significant: a.HasLimitRange != b.HasLimitRange,
		},
	}
	
	// Add top usage comparisons if both have quotas
	if a.HasResourceQuota && b.HasResourceQuota {
		if cpuA, okA := a.TopUsage["cpu"]; okA {
			if cpuB, okB := b.TopUsage["cpu"]; okB {
				rows = append(rows, CompareRow{
					Category:    "Quota",
					Metric:      "CPU Usage",
					ValueA:      cpuA,
					ValueB:      cpuB,
					Severity:    c.calculateUsageSeverity(cpuA, cpuB),
					Trend:       c.calculateTrend(cpuA, cpuB),
					Significant: c.isSignificantUsageDiff(cpuA, cpuB),
				})
			}
		}
		
		if memA, okA := a.TopUsage["memory"]; okA {
			if memB, okB := b.TopUsage["memory"]; okB {
				rows = append(rows, CompareRow{
					Category:    "Quota",
					Metric:      "Memory Usage",
					ValueA:      memA,
					ValueB:      memB,
					Severity:    c.calculateUsageSeverity(memA, memB),
					Trend:       c.calculateTrend(memA, memB),
					Significant: c.isSignificantUsageDiff(memA, memB),
				})
			}
		}
	}
	
	return rows
}

func (c *Comparator) compareStorage(a, b StorageMetrics) []CompareRow {
	rows := []CompareRow{
		{
			Category:    "Storage",
			Metric:      "PVC Count",
			ValueA:      a.PVCCount,
			ValueB:      b.PVCCount,
			Severity:    "normal",
			Trend:       c.calculateIntTrend(a.PVCCount, b.PVCCount),
			Significant: math.Abs(float64(a.PVCCount-b.PVCCount)) > 5,
		},
		{
			Category:    "Storage",
			Metric:      "Requested Storage (Gi)",
			ValueA:      a.RequestedGi,
			ValueB:      b.RequestedGi,
			Severity:    c.calculateStorageSeverity(a.RequestedGi, b.RequestedGi),
			Trend:       c.calculateIntTrend(a.RequestedGi, b.RequestedGi),
			Significant: math.Abs(float64(a.RequestedGi-b.RequestedGi)) > 10,
		},
		{
			Category:    "Storage",
			Metric:      "Unbound PVCs",
			ValueA:      a.UnboundPVCs,
			ValueB:      b.UnboundPVCs,
			Severity:    c.calculateUnboundSeverity(a.UnboundPVCs, b.UnboundPVCs),
			Trend:       c.calculateIntTrend(a.UnboundPVCs, b.UnboundPVCs),
			Significant: a.UnboundPVCs != b.UnboundPVCs,
		},
	}
	
	return rows
}

func (c *Comparator) compareRBAC(a, b RBACMetrics) []CompareRow {
	rows := []CompareRow{
		{
			Category:    "RBAC",
			Metric:      "Admin Bindings",
			ValueA:      a.AdminBindings,
			ValueB:      b.AdminBindings,
			Severity:    c.calculateRBACSeverity(a.AdminBindings, b.AdminBindings),
			Trend:       c.calculateIntTrend(a.AdminBindings, b.AdminBindings),
			Significant: a.AdminBindings != b.AdminBindings,
		},
		{
			Category:    "RBAC",
			Metric:      "Wildcard Rules",
			ValueA:      a.WildcardRules,
			ValueB:      b.WildcardRules,
			Severity:    c.calculateWildcardSeverity(a.WildcardRules, b.WildcardRules),
			Trend:       c.calculateIntTrend(a.WildcardRules, b.WildcardRules),
			Significant: a.WildcardRules != b.WildcardRules,
		},
	}
	
	return rows
}

// Helper functions for severity calculations
func (c *Comparator) calculateResourceSeverity(a, b string) string {
	// Parse resource values and compare
	// Simplified for now
	if a == b {
		return "normal"
	}
	return "warning"
}

func (c *Comparator) calculateHeadroomSeverity(a, b float64) string {
	if a < 20 || b < 20 {
		return "critical"
	}
	if a < 40 || b < 40 {
		return "warning"
	}
	return "normal"
}

func (c *Comparator) calculateStabilitySeverity(a, b int) string {
	if a > 5 || b > 5 {
		return "critical"
	}
	if a > 0 || b > 0 {
		return "warning"
	}
	return "normal"
}

func (c *Comparator) calculatePendingSeverity(a, b int) string {
	if a > 0 || b > 0 {
		return "critical"
	}
	return "normal"
}

func (c *Comparator) calculateExposureSeverity(a, b int) string {
	if a > 10 || b > 10 {
		return "warning"
	}
	return "normal"
}

func (c *Comparator) calculateNetPolSeverity(a, b bool) string {
	if !a || !b {
		return "warning"
	}
	return "normal"
}

func (c *Comparator) calculateQuotaSeverity(a, b bool) string {
	if !a || !b {
		return "warning"
	}
	return "normal"
}

func (c *Comparator) calculateUsageSeverity(a, b string) string {
	// Parse percentage values
	aVal := c.parsePercentage(a)
	bVal := c.parsePercentage(b)
	
	if aVal > 90 || bVal > 90 {
		return "critical"
	}
	if aVal > 70 || bVal > 70 {
		return "warning"
	}
	return "normal"
}

func (c *Comparator) calculateStorageSeverity(a, b int) string {
	if a > 100 || b > 100 {
		return "warning"
	}
	return "normal"
}

func (c *Comparator) calculateUnboundSeverity(a, b int) string {
	if a > 0 || b > 0 {
		return "critical"
	}
	return "normal"
}

func (c *Comparator) calculateRBACSeverity(a, b int) string {
	if a > 5 || b > 5 {
		return "warning"
	}
	return "normal"
}

func (c *Comparator) calculateWildcardSeverity(a, b int) string {
	if a > 0 || b > 0 {
		return "critical"
	}
	return "normal"
}

// Helper functions for trend calculations
func (c *Comparator) calculateTrend(a, b string) string {
	// Simplified comparison
	if a == b {
		return "equal"
	}
	// Would need to parse values properly
	return "up"
}

func (c *Comparator) calculateIntTrend(a, b int) string {
	if a == b {
		return "equal"
	}
	if a < b {
		return "up"
	}
	return "down"
}

func (c *Comparator) calculateFloatTrend(a, b float64) string {
	if math.Abs(a-b) < 0.01 {
		return "equal"
	}
	if a < b {
		return "up"
	}
	return "down"
}

// Helper functions
func (c *Comparator) boolToString(b bool) string {
	if b {
		return "Yes"
	}
	return "No"
}

func (c *Comparator) isSignificantDifference(a, b string, threshold float64) bool {
	// Would need to parse resource values properly
	return a != b
}

func (c *Comparator) isSignificantUsageDiff(a, b string) bool {
	aVal := c.parsePercentage(a)
	bVal := c.parsePercentage(b)
	return math.Abs(aVal-bVal) > 20
}

func (c *Comparator) parsePercentage(s string) float64 {
	// Remove % and parse
	s = strings.TrimSuffix(s, "%")
	val, _ := strconv.ParseFloat(s, 64)
	return val
}